rfcs/build-system/01_convergence_devstack.md

```diff
@@ -0,0 +1,191 @@
+# WIP | FUI Convergence - devstack/DX
+
+> **tags:** pitch, convergence
+
+> For instructions on shaping a project see here: [Shaping a Project](/kX02SXVbS6KzMOQd56i6Cg)
+>
+> Ported from [hackmd](https://hackmd.io/MiS9ISfjRXqmd6XmxpyMaA)
+
+### Problem
+
+Based on [high level repo issues notes](https://hackmd.io/0TIUjU8_T1Ga2Wjez4IiSQ), we need to address most crucial ones sooner than later to enable good enough DX for both inside/outside contributors.
+
+> Good enough DX:
+>
+> - rapid feedback loop
+> - one common way how to do/run things
+> - docs (reflected within tooling -> `yarn test --help` should just work)
+
+### Appetite
+
+As an initial step lets focus on resolve following (smallest possible area):
+
+1. need to build before run/test/demo
+1. unified TSConfigs
+1. unified way how to transpile javascript
+1. solution for convergence/whole repo
+
+### Solution
+
+> show don't tell / TL;DR
+> ![](https://i.imgur.com/bDwWu9v.png)
+
+#### 1. need to build before run/test/lint/demo
+
+Current state:
+
+- N\* uses wepback aliases with lerna-alias
+- OUIF uses yarn workspaces heavily (symlinks of build assets), thus after every modification outside the package that is being worked on, you need to manually rebuild to get latest changes.
+
+Convergence proposal:
+
+- define one source of truth for aliasing packages that can be leveraged consistently in all tooling (bundlers,testing,demos/storybook)
+
+**👉 tsconfig path aliases + TS solution style configs**
+
+Rather then leveraging solutions that might feel appealing (like TS Project refs etc, lerna/yarn workspace symlinks), we can leverage battle tested solution that has been in place for quite some time, thus makes its adoption compatibility with 3rd party toolign straightforward.
```

Comment:
From my understanding this just means that eslint might have slightly slower performance and needs to restart more frequently. This is not good i agree.

But a real world example - the entire Teams codebase is run on TS Project references with eslint and even custom eslint rules in the monorepo. I have never had a problem with using it in vscode and that codebase has smth like 20k modules to build.

Reply:
CI also runs full eslint runs on the entire TMP codebase too on every PR update and merge to master

Reply:
>  slower performance

What means slower performance ? do we have any metrics ? 

> But a real world example...

As can be seen in the official issue, Project refs don't work yet (without enabling experimental flags added recently, which might cause issues or not).

Reply:
It takes ~10 seconds for TS/JS language features to startup in my vscode when I open the teams codebase.

I understand that there is an issue, but the result is that the entire TMP codebase does use eslint with project refs. If you would like to pair up, I'd be happy to dig with you a bit to see what's actually going on there since you have more experience

Reply:
A question I had on TS Project refs and the proposed build setup here: 
- With the current proposal, are we able to easily upgrade to TS Projects later once they get a bit broader support?

This might be a good idea, especially if the current proposal improves the status quo. We don't have to get to the ideal build system in 1 step. It might be a good idea to grab "known working, and good enough for now" to give us room and time to experiment on the future. We should also, however, avoid painting ourselves into a corner 😄